### PR TITLE
Allow using context.Context in middleware and handlers

### PR DESCRIPTION
**Pull request by zserge**
This PR allows getting macaron.Context and models.ReqContext from a standard context.Context (or request.Context()) and is the next step to replacing Macaron.

As a demonstration, it also includes Tracing, Logging and GZip middleware rewritten in a traditional manner.

### CODE REVIEW

File: pkg/macaron/router.go [hunk 1]
```diff
@@ -176,7 +173,7 @@ func (r *Router) Handle(method string, pattern string, handlers []Handler) *Rout
 		h = append(h, handlers...)
 		handlers = h
 	}
-	handlers = validateAndWrapHandlers(handlers, r.handlerWrapper)
```
Comment (marefr) [inline pkg/macaron/router.go:179]: Why is `r.handlerWrapper` not needed any longer and/or what was it used for?

File: pkg/middleware/gziper.go [hunk 2]
```diff
@@ -18,26 +21,62 @@ var gzipIgnoredPathPrefixes = []string{
...
+	grw.Header().Del("Content-Length")
+	return grw.w.Write(p)
+}
+
+func (grw gzipResponseWriter) Hijack() (net.Conn, *bufio.ReadWriter, error) {
+	if hijacker, ok := grw.ResponseWriter.(http.Hijacker); ok {
+		return hijacker.Hijack()
+	}
+	return nil, nil, fmt.Errorf("GZIP ResponseWriter doesn't implement the Hijacker interface")
+}
```
Comment (marefr) [inline pkg/middleware/gziper.go:48]: Maybe can move this into sub-package of macaron for now?

File: pkg/middleware/logger.go [hunk 3]
```diff
@@ -26,29 +26,32 @@ import (
...
-		}
+			if timer, ok := c.Data["perfmon.timer"]; ok {
+				timerTyped := timer.(prometheus.Summary)
+				timerTyped.Observe(float64(timeTaken))
+			}
 
-		status := rw.Status()
-		if status == 200 || status == 304 {
-			if !cfg.RouterLogging {
+			rw := res.(macaron.ResponseWriter)
```
Comment (marefr) [inline pkg/middleware/logger.go:45]: Should probably check that the cast is possible

File: pkg/middleware/request_tracing.go [hunk 4]
```diff
@@ -23,6 +23,8 @@ func ProvideRouteOperationName(name string) macaron.Handler {
 	return func(res http.ResponseWriter, req *http.Request, c *macaron.Context) {
 		ctx := context.WithValue(c.Req.Context(), routeOperationNameKey, name)
 		c.Req.Request = c.Req.WithContext(ctx)
+		// XXX (zserge) Can't find a better way to propagate modified request to the previous middleware in the chain
+		*req = *c.Req.Request
```
Comment (marefr) [inline pkg/middleware/request_tracing.go:27]: Yeah this looks a bit weird. But :shrug:

File: pkg/middleware/gziper.go [hunk 5]
```diff
@@ -18,26 +21,62 @@ var gzipIgnoredPathPrefixes = []string{
...
+	macaron.ResponseWriter
+}
+
+func (grw *gzipResponseWriter) WriteHeader(c int) {
+	grw.Header().Del("Content-Length")
+	grw.ResponseWriter.WriteHeader(c)
+}
+
+func (grw gzipResponseWriter) Write(p []byte) (int, error) {
+	fmt.Println("GZIP WRITE", grw.Header())
```
Comment (marefr) [inline pkg/middleware/gziper.go:35]: Remove?

**Comment by zserge** [comment]
Closing this PR, as it has been replaced by smaller PRs like #37795 #37874 #37672 #37627 #37194 #37057 etc

**Pull request closed without merging**